In [1]:
import pandas as pd


In [2]:
df = pd.read_csv("messy_sales.csv")

In [3]:
df.shape

(300, 6)

In [4]:
df.head(10)

,order_id,date,product,price,qty,zip
0,1254,04/06/2026,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,05/06/2026,mug,5.50,1,10001
3,1066,05/30/2026,notebook,NaN,1,98101
4,1114,04/06/2026,webcam,45.59,4,90405
5,1280,04/22/2026,pen set,NaN,4,30303
6,1204,04/18/2026,charger,8.41,2,2134
7,1249,2026-05-15,keyboard,75.04,4,90405
8,1037,2026-05-19,mug,9.31,2,60614
9,1177,2026-05-13,mug,10.11,3,90210


In [5]:
df.dtypes



order_id      int64
date            str
product         str
price       float64
qty           int64
zip           int64
dtype: object

# Initial Data Problems

1. Some ZIP codes have only 4 digits, so a leading zero may be missing.
2. The dates are recorded in different formats.
3. Some prices show NaN, which means the price value is missing.

In [6]:
df.isna().sum()

order_id     0
date         0
product      0
price       12
qty          0
zip          0
dtype: int64

In [7]:
df["price"].median()

np.float64(37.53)

In [8]:
fill_value = df["price"].median()
print(fill_value)

37.53


In [9]:
df["price"] = df["price"].fillna(fill_value)


In [10]:
df.isna().sum()

order_id    0
date        0
product     0
price       0
qty         0
zip         0
dtype: int64

In [11]:
df.duplicated().sum()

np.int64(8)

In [12]:
df = df.drop_duplicates()

In [13]:
df.shape


(292, 6)

In [14]:
df["zip"] = df["zip"].astype(str).str.zfill(5)

In [15]:
df["zip"].head(10)

0    60614
1    30303
2    10001
3    98101
4    90405
5    30303
6    02134
7    90405
8    60614
9    90210
Name: zip, dtype: str

In [16]:
df["date"] = pd.to_datetime(df["date"], format="mixed")

In [17]:
df.dtypes

order_id             int64
date        datetime64[us]
product                str
price              float64
qty                  int64
zip                    str
dtype: object

In [18]:
df.shape

(292, 6)

In [19]:
df[df["qty"] < 0]

,order_id,date,product,price,qty,zip
202,1140,2026-04-18,webcam,38.69,-5,98101
262,1233,2026-05-09,desk lamp,22.64,-4,10001
297,1025,2026-04-27,keyboard,47.30,-2,02116


# Negative Quantity Decision

I saw 3 rows with negative quantities. I desided  tho leave these rows because there is not enough information to conclude that they are data-entry errors. Probably, these negative quantities may represent product returns, and removing or changing them could incorrectly increase the reported net sales quantity.

In [20]:
df.to_csv("sales_clean.csv", index=False)

# Cleaning Log

Loaded 300 rows and 6 columns.
Found 12 missing price values and filled them with the median price of $37.53,because the median is resistant to skew
Removed 8 duplicate rows. Now the total is 292 rows.
Converted ZIP codes from numbers to text and restored leading zeros so that all ZIP codes contain 5 digits.
Standardized the date format.
Found 3 rows with negative quantities and kept them because there was not enough information to conclude that they were data-entry errors.
Final dataset contains 292 rows and 6 columns.

Wrong cleaning decisions could change the sales results and lead to incorrect conclusions about the data.
